File: challenges/xlang/cs/spectralnorm.cs
using System;
using System.Globalization;

static class SpectralNorm
{
    static double A(int i, int j) => 1.0 / ((i + j) * (i + j + 1) / 2 + i + 1);

    static void MulAv(double[] v, double[] outv)
    {
        int n = v.Length;
        for (int i = 0; i < n; i++) { double s = 0; for (int j = 0; j < n; j++) s += A(i, j) * v[j]; outv[i] = s; }
    }
    static void MulAtv(double[] v, double[] outv)
    {
        int n = v.Length;
        for (int i = 0; i < n; i++) { double s = 0; for (int j = 0; j < n; j++) s += A(j, i) * v[j]; outv[i] = s; }
    }
    static void MulAtAv(double[] v, double[] outv, double[] tmp) { MulAv(v, tmp); MulAtv(tmp, outv); }

    static int Main(string[] args)
    {
        int n = args.Length > 0 && int.TryParse(args[0], out int x) ? x : 100;
        double[] u = new double[n]; double[] v = new double[n]; double[] tmp = new double[n];
        for (int i = 0; i < n; i++) u[i] = 1.0;
        for (int i = 0; i < 10; i++) { MulAtAv(u, v, tmp); MulAtAv(v, u, tmp); }
        double vbv = 0, vv = 0;
        for (int i = 0; i < n; i++) { vbv += u[i] * v[i]; vv += v[i] * v[i]; }
        Console.WriteLine(Math.Sqrt(vbv / vv).ToString("F9", CultureInfo.InvariantCulture));
        return 0;
    }
}

File: challenges/xlang/rs/spectralnorm.rs
fn a(i: usize, j: usize) -> f64 { 1.0 / (((i + j) * (i + j + 1) / 2 + i + 1) as f64) }

fn mul_av(v: &[f64], out: &mut [f64]) {
    let n = v.len();
    for i in 0..n { let mut s = 0.0; for j in 0..n { s += a(i, j) * v[j]; } out[i] = s; }
}
fn mul_atv(v: &[f64], out: &mut [f64]) {
    let n = v.len();
    for i in 0..n { let mut s = 0.0; for j in 0..n { s += a(j, i) * v[j]; } out[i] = s; }
}
fn mul_atav(v: &[f64], out: &mut [f64], tmp: &mut [f64]) { mul_av(v, tmp); mul_atv(tmp, out); }

fn main() {
    let n: usize = std::env::args().nth(1).and_then(|a| a.parse().ok()).unwrap_or(100);
    let mut u = vec![1.0f64; n];
    let mut v = vec![0.0f64; n];
    let mut tmp = vec![0.0f64; n];
    for _ in 0..10 { mul_atav(&u, &mut v, &mut tmp); mul_atav(&v, &mut u, &mut tmp); }
    let mut vbv = 0.0; let mut vv = 0.0;
    for i in 0..n { vbv += u[i] * v[i]; vv += v[i] * v[i]; }
    println!("{:.9}", (vbv / vv).sqrt());
}
